core/retries-spi/src/main/java/software/amazon/awssdk/retries/api/RetryStrategy.java

```diff
@@ -198,12 +198,24 @@ default B retryOnRootCauseInstanceOf(Class<? extends Throwable> throwable) {
         B maxAttempts(int maxAttempts);
 
         /**
-         * Configure the backoff strategy used by this executor.
+         * Configure the backoff strategy used by the retry strategy.
          *
          * <p>By default, this uses jittered exponential backoff.
          */
         B backoffStrategy(BackoffStrategy backoffStrategy);
 
+        /**
+         * Configure the backoff strategy used for throttling exceptions by this strategy.
+         *
+         * <p>By default, this uses jittered exponential backoff.
+         */
+        B throttlingBackoffStrategy(BackoffStrategy throttlingBackoffStrategy);
+
+        /**
+         * Configure the predicate to allow the strategy categorize a Throwable as throttling exception.
+         */
+        B treatAsThrottling(Predicate<Throwable> treatAsThrottling);
```

Comment:
I'm worried about baking the concept of "throttling exceptions" into our base retry strategy API. Is it a better or worse API if we allow them to specify a `Function<Throwable, BackoffStrategy>` for having an exception-level override for backoff strategies for any reason, instead of making this API specific to throttling exceptions?

i.e. a "use this backoff strategy for this exception type, overriding the default".  We'd use it for throttling exceptions:

```
throwable -> isThrottlingException(throwable) ? throttlingBackoffStrategy : null; // null = use default  backoff strategy
```

But it would be more flexible because it could be used to override the backoff strategy for IOExceptions or DynamoDbExceptions, etc.

Reply:
I'm imagining just one new API instead of two:

```java
/**
 * Specify a function that overrides the backoff strategy for a particular exception type. When a throwable is considered retryable, the SDK will use the default backoff strategy unless this function returns a different one to use.
 */
B overrideBackoffStrategy(Function<Throwable, BackoffStrategy> overrideBackoffStrategyFunction);
```

Reply:
Discussed offline. We already have methods specific to throttling in adaptive and legacy, so this is a door we've mostly already walked through. We'll use these two methods everywhere and if we ever want to make it more flexible, we can do that then instead of trying to do it now.